org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionNodeDetector.java

```diff
@@ -522,14 +522,30 @@ private void endVisit(ASTNode astNode) {
 					&& astNode != this.searchedNode) {
 				if(!(astNode instanceof AllocationExpression && ((AllocationExpression) astNode).type == this.searchedNode)
 					&& !(astNode instanceof ConditionalExpression && ((ConditionalExpression) astNode).valueIfTrue == this.searchedNode)
-					&& !(astNode instanceof ConditionalExpression && ((ConditionalExpression) astNode).valueIfFalse == this.searchedNode)) {
+					&& !(astNode instanceof ConditionalExpression && ((ConditionalExpression) astNode).valueIfFalse == this.searchedNode)
+					// if the completion is on a case statement, accept the switch as parent.
+					&& (!(astNode instanceof SwitchStatement ss) || isOnCompletingOnCaseLabel(ss))) {
```

Comment:
Following the style of #1953 we could move this into `endVisit(SwitchStatement,BlockScope)`  like this:
```
		endVisit(switchStatement);
		if (this.parent == switchStatement && !isOnCompletingOnCaseLabel(switchStatement))
			this.parent = NOT_A_PARENT;
```

I'd prefer that over bloating the current method. The only downside: we'd have to repeat those 2 lines also in `endVisit(SwitchExpression,BlockScope)`.

WDYT?